src/bam/mod.rs

```diff
@@ -403,6 +403,157 @@ impl Drop for Writer {
 }
 
 
+// ---- SAMWriter ------------------
+
+/// SAM writer.
+pub struct SAMWriter {
+    f: *mut htslib::htsFile,
+    header: HeaderView,
+}
+
+/// Wrapper for opening a SAM file.
+fn hts_open(path: &ffi::CStr, mode: &[u8]) -> Result<*mut htslib::htsFile, SAMError> {
+    let ret = unsafe {
+        htslib::hts_open(
+            path.as_ptr(),
+            ffi::CString::new(mode).unwrap().as_ptr()
+        )
+    };
+    if ret.is_null() {
+        Err(SAMError::IOError)
+    } else {
+        Ok(ret)
+    }
+}
+
+
+impl SAMWriter {
+    /// Create new SAM file writer.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - the path.
+    /// * `header` - header definition to use
+    pub fn from_path<P: AsRef<Path>>(path: P, header: &header::Header) -> Result<Self, SAMError> {
+        if let Some(p) = path.as_ref().to_str() {
+            Ok(try!(Self::new(p.as_bytes(), header)))
+        } else {
+            Err(SAMError::IOError)
+        }
+    }
+
+    /// Create a new SAM file at STDOUT.
+    ///
+    /// # Arguments
+    ///
+    /// * `header` - header definition to use
+    pub fn from_stdout(header: &header::Header) -> Result<Self, SAMError> {
+        Self::new(b"-", header)
+    }
+
+    fn new(path: &[u8], header: &header::Header) -> Result<Self, SAMError> {
```

Comment:
There is quite some code duplication here. Maybe BAM and SAM can share a common trait?